Method to make multiple animations in one figure

johnomotani opened the pull request:
Plot multiple variables at once, like
```
ds.bout.animate_list([n, V, U])
```
The variables need to be sliced to 1D+1 or 2D+1 first (can mix 1d and 2d in one call though).

pep8speaks commented:
Hello @johnomotani! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`xbout/boutdataset.py`](https://github.com/boutproject/xBOUT/blob/9265ceedc8926b60588ac497dfd5c9535867ddca/xbout/boutdataset.py):

> [Line 246:13](https://github.com/boutproject/xBOUT/blob/9265ceedc8926b60588ac497dfd5c9535867ddca/xbout/boutdataset.py#L246): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 270:21](https://github.com/boutproject/xBOUT/blob/9265ceedc8926b60588ac497dfd5c9535867ddca/xbout/boutdataset.py#L270): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 273:21](https://github.com/boutproject/xBOUT/blob/9265ceedc8926b60588ac497dfd5c9535867ddca/xbout/boutdataset.py#L273): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 276:25](https://github.com/boutproject/xBOUT/blob/9265ceedc8926b60588ac497dfd5c9535867ddca/xbout/boutdataset.py#L276): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 276:26](https://github.com/boutproject/xBOUT/blob/9265ceedc8926b60588ac497dfd5c9535867ddca/xbout/boutdataset.py#L276): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator

johnomotani commented:
Sorry, I made this PR too soon, should have tried out the new method more first. I've rebased on the current version of 1d-animations now, and fixed a few issues with the original code at the same time.

TomNicholas commented on xbout/boutdataset.py:
```diff
@@ -239,6 +242,50 @@ def to_restart(self, savepath='.', nxpe=None, nype=None,
+        elif ncols is None:
+            ncols = int(np.ceil(nvars/nrows))
+        else:
+            assert nrows*ncols >= nvars, 'nrows*ncols not as big as nvar'
```
Asserts [shouldn't be used for error/warning checking](https://stackoverflow.com/questions/944592/best-practice-for-python-assert) (or for any flow control in general), because they will be [ignored](https://stackoverflow.com/questions/2830358/what-are-the-implications-of-running-python-with-the-optimize-flag.
) if you run python with the `-O` or `-OO` optimisation flags.

This should instead be
```python
if nrows * ncols < nvars:
    raise ValueError("Not enough rows * columns to fit all variables")
```

TomNicholas commented on xbout/boutdataset.py:
```diff
@@ -239,6 +242,50 @@ def to_restart(self, savepath='.', nxpe=None, nype=None,
+            ax.set_title(data.name)
+
+            if ndims == 2:
+                blocks.append(animate_line(data=data, ax=ax, animate_over=animate_over,
```
This is really neat, and a nice example of the usefulness of `animatplot`'s `block` abstraction.

TomNicholas commented on xbout/plotting/animate.py:
```diff
@@ -7,7 +7,7 @@
 
 
 def animate_imshow(data, animate_over='t', x='x', y='y', animate=True,
```
It's likely that we shouldn't actually be using `imshow`, because it's restricted to square images right? Though I originally wrote this for slab cross-sections, then really this should use `pcolormesh` so that it can handle any geometry? Integrating it with the multi-region plotting method used in #34 will add a bit of complexity (though I think the blocks abstraction will mean it's not too hard).

johnomotani replied:
Think I agree with all of this, although it does need a bit of thinking... Having poloidal plots with geometry included is (very, very) nice to have as an option (maybe even default), but having cartesian as an option too is also useful - for example poloidal plotting of some arbitrary slice is going to be a bit tricky*; also if there are some very small or distorted cells they may be easier to see in a plot that's just rectangular. The blocks abstraction does very much help with all of this, we can add over time functions that create a block in various ways, and code like this that sticks blocks together would only need to add some logic to say which function to call.

It probably would be sensible to use `pcolormesh` instead of `imshow` everywhere though, since it's much more flexible.

--
* Maybe it would be useful to introduce some 'coordinates' that say which region a point is, e.g. `xregion` where `xregion=0` for `x<ixseps1`, `xregion=1` for `ixseps1<=x<ixseps2`, `xregion=2` for `x>=ixseps2` [note ixseps1/ixseps2 include the x-boundary cells, so need to subtract `MXG` if x does not include the boundary cells]. Hopefully this can be done in such a way that if you do something like `n.isel(x=slice(3,7,1)` then `xregion` would get sliced as well. Then we could select regions to join with `xregion` and a `yregion` (defined using `jyseps*`)?

johnomotani commented:
> This is cool. Have you tried running it with real data? I'm a bit concerned that there might be bugs upstream in animatplot.imshow. Also we should think about how the 2D animations will combine with #34.

I have been using it, and it works. Animations sometimes seem to freeze in a jupyter notebook, but I've not worked out how to reproduce yet. It might also be an issue with my setup, I'm currently having problems with matplotlib in ipython, where pyplot.show() is not blocking. Need to spend more time observing when I have problems before trying seriously to debug...

johnomotani commented:
This PR now includes most of the work I'd done on my dev branch:
* uses `pcolormesh` instead of `imshow` for animations
* enables animations of poloidal-plane `(R,Z)` plots
* various small fixes/features

johnomotani commented on xbout/plotting/utils.py:
```diff
@@ -41,26 +41,29 @@ def _decompose_regions(da):
     j11, j12, j21, j22, ix1, ix2, nin, _, ny, y_boundary_guards = _get_seps(da)
     regions = []
 
+    x, y = da.dims[-2:]
```
Might be able to do this better, without relying on order of dimensions, if #68 is merged.

rdoyle45 commented on xbout/plotting/animate.py:
```diff
@@ -1,21 +1,129 @@
+
+        if not save_as:
+            save_as = "{}_over_{}".format(da.name, animate_over)
+        anim.save(save_as + '.gif', writer='imagemagick')
```
We should probably use `PillowWriter(fps=fps)` as in the other animate functions

rdoyle45 commented on xbout/plotting/animate.py:
```diff
@@ -1,21 +1,129 @@
+    return blocks
+
 from .utils import plot_separatrix
 from matplotlib.animation import PillowWriter
```
Can this be moved to the top of the script?

rdoyle45 replied:
Can this be moved to the top of the script?

rdoyle45 commented on xbout/plotting/animate.py:
```diff
@@ -1,21 +1,129 @@
+
+    return blocks
+
 from .utils import plot_separatrix
```
This can be removed, as its already above

rdoyle45 commented on xbout/plotting/animate.py:
```diff
@@ -1,21 +1,129 @@
+
+    return blocks
+
 from .utils import plot_separatrix
```
`plot_separatrix` is no longer used?

johnomotani replied:
`plot_separatrix` isn't used in the library, but does a different thing from `plot_separatrices` - it plots a separatrix at a given x-position, and can be used with x-z plots, while `plot_separatrices` works on R-Z plots. It can usefully be called by the user to add a separatrix. It might be good to make the naming of the two functions less confusing, and possibly make `plot_separatrix` get the separatrix positions from `ixseps1` and `ixseps2` (if available).

rdoyle45 replied:
It used to be used by `animate_imshow`, I just meant it doesn't need to imported anymore. This was solved by https://github.com/boutproject/xBOUT/pull/38/commits/85b0cc6ebe816604b5994662e99c5c33ac35d6f0 .

rdoyle45 commented:
Looks good to me. Did a few animations, no issues, worked well.

TomNicholas commented on xbout/plotting/plotfuncs.py:
```diff
@@ -57,8 +61,20 @@ def plot2d_wrapper(da, method, *, ax=None, separatrix=True, targets=True,
         Draw hatched areas at the targets
+    gridlines : bool or int, optional
+        If True, draw grid lines on the plot. If an int is passed, it is used as the
+        stride when plotting grid lines (to reduce the number on the plot)
```
It would have been good to separate these into a different PR but no matter

TomNicholas commented on xbout/plotting/plotfuncs.py:
```diff
@@ -169,6 +196,24 @@ def plot2d_wrapper(da, method, *, ax=None, separatrix=True, targets=True,
+    if gridlines is not None:
+        if gridlines is True:
+            gridlines = (1, 1)
+        if not isinstance(gridlines, collections.Sequence):
```
`collections.sequence` seems an odd choice of container class? We just need to store either one or two ints right? Why not just a short list? Or even a dictionary?
```python
gridlines= {'x': 1, 'y':1}
```

johnomotani replied:
Sorry, this was written quickly. Restricting to `int`, `slice` or `dict of int or slice` is better.

TomNicholas commented on xbout/plotting/plotfuncs.py:
```diff
@@ -169,6 +196,24 @@ def plot2d_wrapper(da, method, *, ax=None, separatrix=True, targets=True,
+        Z_regions = _decompose_regions(da['Z'])
+
+        for R, Z in zip(R_regions, Z_regions):
+            plt.plot(R[::gridlines[0], :].T, Z[::gridlines[0], :].T, color='k', lw=0.1)
```
Using a dictionary would eliminate the "magic numbers" 0 and 1 here

TomNicholas commented on xbout/tests/test_animate.py:
```diff
@@ -58,3 +60,228 @@ def test_animate1D(self, create_test_file):
+        assert isinstance(animations.blocks[1], Pcolormesh)
+        assert isinstance(animations.blocks[2], Line)
+
+    def test_animate_list_fps(self, create_test_file):
```
This isn't testing that the fps was actually set. Should do 
```python
assert animations.fps == fps
```

TomNicholas replied:
(also to be really nitpicky the animatplot logic is that you have created one animation (singular) which consists of multiple blocks (plural))

johnomotani commented:
@TomNicholas 
> This is a big PR!

It is, sorry! Do you want me to try and split it up?

TomNicholas commented:
> It is, sorry! Do you want me to try and split it up?

That might be a good idea - there is at least one clear division: the new arguments to `plot2d_wrapper` can be their own PR (so gridlines etc. can be discussed elsewhere).

johnomotani commented on xbout/boutdataarray.py:
```diff
@@ -75,14 +83,31 @@ def animate2D(self, animate_over='t', x=None, y=None, animate=True,
+                print("{} data passed has {} dimensions - making poloidal plot with "
+                      "animate_poloidal()".format(variable, str(n_dims)))
+                if x is not None:
+                    kwargs['x'] = x
```
I don't think so. Leaving the default as `None` passes off setting defaults to the called methods, so lets `animate_poloidal` use `(R,Z)` and `animate_pcolormesh` use `da.dims`.

johnomotani commented on xbout/boutdataset.py:
```diff
@@ -158,6 +165,149 @@ def to_restart(self, savepath='.', nxpe=None, nype=None,
+                raise ValueError('if poloidal_plot is a sequence, it must have the same '
+                                 'number of elements as "variables"')
+        except TypeError:
+            poloidal_plot = [poloidal_plot] * len(variables)
```
Just as somewhere to note this: to test for a sequence, as suggested on https://stackoverflow.com/questions/2937114/python-check-if-an-object-is-a-sequence, probably best pattern is to use `isinstance(variable, collections.Sequence)`.

johnomotani commented:
Made some updates to address @TomNicholas's review. Will try to split PR now.

johnomotani commented:
Replaced by #71, #72 and #73.
